src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,1022 @@
+bit_range_mut_display!(u64, NonZeroU64);
+bit_range_mut_display!(u32, NonZeroU32);
+bit_range_mut_display!(u16, NonZeroU16);
+bit_range_mut_display!(u8, NonZeroU8);
+
+/// Macro for defining `impl BitRange`.
+macro_rules! bit_range {
+    ($x:ty, $y:ty, $mask:ident, $max:ident) => {
+        impl<const START: u8, const END: u8> BitRange<'_, $x, START, END> {
+            pub const MASK: $x = $mask::<START, END>();
+            /// The maximum value this range can store
+            pub const MAX: $x = $max::<START, END>();
+
+            #[doc = concat!("
+                Returns the value of the bit range.
+                ```
+                use bit_fields::BitRange;
+                let x = 18", stringify!($x), ";
+                assert_eq!(BitRange::<_,0,4>(&x),2);
+                assert_eq!(BitRange::<_,4,8>(&x),1);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub fn read(&self) -> $x {
+                <$x>::from(self)
+            }
+
+            #[doc = concat!("
+                Convenience alias for [`Self::MAX`].
+                
+                ```
+                use bit_fields::BitRange;
+                let x = ", stringify!($x), "::default();
+
+                assert_eq!(BitRange::<_,0,1>(&x).get_max(),1);
+                assert_eq!(BitRange::<_,0,2>(&x).get_max(),3);
+                assert_eq!(BitRange::<_,0,3>(&x).get_max(),7);
+                assert_eq!(BitRange::<_,0,4>(&x).get_max(),15);
+                assert_eq!(BitRange::<_,0,5>(&x).get_max(),31);
+                assert_eq!(BitRange::<_,0,6>(&x).get_max(),63);
+                assert_eq!(BitRange::<_,0,7>(&x).get_max(),127);
+                assert_eq!(BitRange::<_,0,8>(&x).get_max(),255);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_max(&self) -> $x {
+                Self::MAX
+            }
+            #[doc = concat!("
+                Convenience alias for [`Self::MASK`].
+                
+                ```
+                use bit_fields::BitRange;
+                let x = ", stringify!($x), "::default();
+
+                assert_eq!(BitRange::<_,0,1>(&x).get_mask(),1);
+                assert_eq!(BitRange::<_,1,3>(&x).get_mask(),6);
+                assert_eq!(BitRange::<_,3,7>(&x).get_mask(),120);
+                assert_eq!(BitRange::<_,7,8>(&x).get_mask(),128);
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_mask(&self) -> $x {
+                Self::MASK
+            }
+        }
+        impl<const START: u8, const END: u8> BitRange<'_, $y, START, END> {
+            // SAFETY: Value known to be non-zero at compile time.
+            pub const MASK: $y = unsafe { <$y>::new_unchecked($mask::<START, END>()) };
+            // SAFETY: Value known to be non-zero at compile time.
+            /// The maximum value this range can store
```

Comment:
nit(safety comment): the compiler does ensure no invalid NonZero* are constructed here, but not because it is known that `$mask::<START, END>()` cannot return zero (it can, e.g. if one tries to construct a `BitRange<_, _, 0, 0>`). The compiler simply error out it can prove that the value passed to new_unchecked is zero at compile time.

Reply:
Adjusted comment:
```rust
// SAFETY: A zero value results in an error at compile time.
```